### PR TITLE
 Add util to dump out BM25 vectors for corpus

### PR DESCRIPTION
**Pull request by manveertamber**
https://github.com/castorini/pyserini/issues/1050

### CODE REVIEW

File: pyserini/index/lucene/_base.py [hunk 1]
```diff
@@ -535,3 +537,36 @@ def stats(self) -> Dict[str, int]:
             index_stats_dict[term] = index_stats_map.get(term)
 
         return index_stats_dict
+
+    def dump_documents_BM25(self, file_path="collections/documents_bm25_dump.json"):
```
Comment (lintool) [inline pyserini/index/lucene/_base.py:541]: Few comments here:
- I don't think output file should have a default value; i.e., the caller should explicitly pass something in.
- please expose the b1 and k values from the underlying method, but keep the standard defaults of 0.9 and 0.4

File: pyserini/index/lucene/_base.py [hunk 2]
```diff
@@ -535,3 +537,36 @@ def stats(self) -> Dict[str, int]:
...
+        """Dumps out all the document vectors with BM25 weights in Pyserini's JSON vector format 
+
+        Parameters
+        ----------
+        file_path : str
+            file path to dump JSON file
+        """
+        first_iter = True
+        output_string =  "[\n"
```
Comment (lintool) [inline pyserini/index/lucene/_base.py:550]: Let's use a real JSON writer so we can programmatically control indent, etc.?

File: pyserini/index/lucene/_base.py [hunk 3]
```diff
@@ -535,3 +537,36 @@ def stats(self) -> Dict[str, int]:
...
+
+        for i in tqdm(range(self.stats()["documents"])):
```
Comment (lintool) [inline pyserini/index/lucene/_base.py:552]: defensive programming here. check 'documents' first to make sure it's not None, etc.

File: pyserini/index/lucene/_base.py [hunk 4]
```diff
@@ -535,3 +537,36 @@ def stats(self) -> Dict[str, int]:
...
+            bm25_vector = {term: self.compute_bm25_term_weight(docid, term, analyzer=None) for term in tf.keys()}
+            
+            if not first_iter:
+                output_string += "  },\n"
+            else:
+                first_iter = False
+            
+            output_string += "  {\n"
+            output_string += "    \"id\": \"" + str(docid) +"\",\n"
+            output_string += "    \"contents\": \"BM25 vector for document: " + str(docid) + "\",\n"
```
Comment (lintool) [inline pyserini/index/lucene/_base.py:564]: Don't think we need to include contents.

File: tests/test_index_reader.py [hunk 5]
```diff
@@ -396,6 +397,38 @@ def test_jstring_encoding(self):
         # the discussion here: https://github.com/castorini/pyserini/issues/770
         JString('zoölogy')
 
+    def test_dump_documents_BM25(self):
```
Comment (lintool) [inline tests/test_index_reader.py:400]: can you please add a test case to do searching?

File: docs/usage-indexreader.md [hunk 6]
```diff
@@ -165,3 +165,29 @@ Output is something like this:
...
 Note that unless the underlying index was built with the `-optimize` option (i.e., merging all index segments into a single segment), `unique_terms` will show -1.
 Nope, that's not a bug.
+
+## How do I dump out BM25 vectors for every document?
+
+Here's how to dump out all the document vectors with BM25 weights in Pyserini's JSON vector format:
+
+```python
+# You can specify a file path for the .json file. Default is "collections/documents_bm25_dump.json"
+index_reader.dump_documents_BM25("collections/cacm_documents_bm25_dump.json")
```
Comment (lintool) [inline docs/usage-indexreader.md:175]: we use single quotes throughout code base, please change.

**Review by lintool — CHANGES_REQUESTED** [review]

File: pyserini/index/lucene/_base.py (hunk 3 above)
Comment (manveertamber) [reply pyserini/index/lucene/_base.py]: Does it look good now?`

File: docs/usage-indexreader.md (hunk 6 above)
Comment (lintool) [reply docs/usage-indexreader.md:175]: change double to single quotes?

File: tests/test_index_reader.py [hunk 7]
```diff
@@ -396,6 +398,49 @@ def test_jstring_encoding(self):
...
+            for term in query_terms:
+                if term in doc['vector']:
+                    score += doc['vector'][term]
+            heapq.heappush(heap, (-1*score, doc['id'])) 
+        # Using LuceneSearcher instead
+        hits = self.searcher.search(query)
+        for i in range(0, 10):
+            top = heapq.heappop(heap)
+            self.assertEqual(hits[i].docid, top[1])
+            self.assertAlmostEqual(hits[i].score, -1*top[0],places=3)
```
Comment (lintool) [inline tests/test_index_reader.py:425]: pedantic nit, but space before 'places'.

Set up your code editor to de-lint for PEP8 compliance? should catch these.

File: tests/test_index_reader.py [hunk 8]
```diff
@@ -396,6 +398,49 @@ def test_jstring_encoding(self):
...
+    def test_dump_documents_BM25(self):
+        filepath = 'collections/cacm_documents_bm25_dump.json'
+        self.index_reader.dump_documents_BM25(filepath)
+        f = open(filepath, 'r')
+        dump = json.load(f)
+        f.close()
+
+        assert(len(dump) == self.index_reader.stats()['documents'])
+
+        query = 'Where can I learn to fly a helicopter?'
```
Comment (lintool) [inline tests/test_index_reader.py:410]: Where did you get these queries from?

These are topics used in the actual CACM collection:
https://github.com/castorini/anserini-data/blob/master/CACM/topics.cacm.txt

Can you change to a few queries selected from this list? Would be a better topical fit...

File: tests/test_index_reader.py [hunk 9]
```diff
@@ -396,6 +398,49 @@ def test_jstring_encoding(self):
...
+            heapq.heappush(heap, (-1*score, doc['id'])) 
+        # Using LuceneSearcher instead
+        hits = self.searcher.search(query)
+        for i in range(0, 10):
+            top = heapq.heappop(heap)
+            self.assertEqual(hits[i].docid, top[1])
+            self.assertAlmostEqual(hits[i].score, -1*top[0],places=3)
+
+        query = 'What is the capital of Canada? In what province is it located? How can i get there from Toronto, Ontario, Canada?'
+        query_terms = self.index_reader.analyze(query, analyzer=analysis.get_lucene_analyzer())
```
Comment (lintool) [inline tests/test_index_reader.py:428]: instead of duplicating code, why not write a helper method and just call with different queries?

**Review by lintool — CHANGES_REQUESTED** [review]
Few more comments...

**Review by lintool — APPROVED** [review]

**Pull request merged**